Add scala.compiletime.requireConst

nicolasstucki opened the pull request:
`requireConst` checks at compiletime that the provided values is a constant after inlining and constant folding.

Usage:
```scala
inline def twice(inline n: Int): Int =
  requireConst(n) // static assertion that the parameter `n` is a constant
  n + n

twice(1)
val m: Int = ...
twice(m) // error: expected a constant value but found: m
```

nicolasstucki commented:
This is an ability we lost when we fixed the semantics of inline parameters. Currently, we can only do these checks using macros.

nicolasstucki commented:
There is also a new question to ask ourselves. Should we drop the `inline if` syntax? `const` trivially subsume all use-cases of `inline if` with no extra syntactic overhead

```diff
- inline if n == 0 then ... else ...
+ if const(n == 0) then ... else ...
```

Using const has also the advantage of linking to some documentation in the IDE and is more consistent with the other compile-time operations such as `summonInline`.

smarter commented:
>   const(n) // static assertion that the parameter `n` is a constant

I find it weird that the same syntax is used in if as a condition and as an assertion. I think the assertion should be more explicit, like `requireConst(n)`

nicolasstucki commented:
There is only one method, that happens to be used in both use cases. If I rename it we would end up with 
```scala
  if requireConst(n == 0) then 1L
  else if requireConst(n % 2 == 1) then x * power(x, n - 1)
```

it is a bit too long to write these inline.

sjrd commented:
I think that `requireConst` is better than `const`. I had shared the same misunderstanding as @smarter, so clearly this is not an isolated problem. If we, who are very used to imagine what it could mean for a compiler, were confused by this in the same way, it's quite likely that non-compiler-developers will be even more confused.

The fact that it's bit longer should not stand in the way of clarity.

nicolasstucki commented:
Ok. I changed `const` to `requireConst`.

smarter commented:
I still find it weird that the same method call does something different (return a value or emitting an error) depending on where it's used syntactically. I don't think we've ever had something like that in Scala.

sjrd commented:
That's not what's happening here. In both cases it a) emits an error if the argument is not constant, and b) returns the constant value of the argument if it is constant.

In the first example, the returned value is just discarded.

smarter commented:
Ah I see, I misread `if const(foo)` and `if requireConst(foo)` as "if foo is const" which is maybe a sign this isn't the best syntax.

nicolasstucki commented:
The doc should explicitly state that the value is returned.

nicolasstucki commented:
Alternative names
* requiredConst
* assertedConst

These names seem to indicate that the value is returned

nicolasstucki commented:
@liufengyun I updated this PR to match what we agreed during the meeting. Now `requireConst` returns `Unit` and I removed the special handling in the `inline if`. I also made the argument type more precise. Unlike other intrinsic macros, there is only a check done but the code generation is just the normal inlining of the method.
```scala
inline def requireConst(inline x: Boolean | Byte | Short | Int | Long | Float | Double | Char | String): Unit = ()
```

liufengyun commented on library/src/scala/compiletime/package.scala:
```diff
@@ -50,6 +50,22 @@ package object compiletime {
   end extension
 
+  /** Checks at compiletime that the provided values is a constant after
+   *  inlining and constant folding and returns the value.
```
```suggestion
   *  inlining and constant folding.
```

liufengyun commented on library/src/scala/compiletime/package.scala:
````diff
@@ -50,6 +50,22 @@ package object compiletime {
+   *  Usage:
+   *  ```scala
+   *  inline def twice(inline n: Int): Int =
+   *    requireConst(n) // static assertion that the parameter `n` is a constant
````
```suggestion
   *    requireConst(n) // compile-time assertion that the parameter `n` is a constant
```

liufengyun reviewed (approved):
LGTM

Side note: I find this is a very good practice to express pre-conditions in meta-programming. 
I can foresee similar APIs will be helpful in macros as well. It helps simplify error handling code and makes the assumptions clear.
